File: src/python/aerocapture/training/rl/layers/lstm.py
"""LSTM cell matching nn.LSTMCell + Rust LstmLayer bit-for-bit.

Canonical flat weight order (LayerWeights trait + PSO chromosome):
    weight_ih row-major, weight_hh row-major, bias_ih, bias_hh.

Forward equations (PyTorch nn.LSTMCell convention, gate order i, f, g, o):
    i_t = sigmoid(W_ii @ x + b_ii + W_hi @ h + b_hi)
    f_t = sigmoid(W_if @ x + b_if + W_hf @ h + b_hf)
    g_t =    tanh(W_ig @ x + b_ig + W_hg @ h + b_hg)
    o_t = sigmoid(W_io @ x + b_io + W_ho @ h + b_ho)
    c_t = f_t * c_{t-1} + i_t * g_t
    h_t = o_t * tanh(c_t)
"""

from __future__ import annotations

from typing import Any

import numpy as np
import torch
from torch import Tensor, nn


class LstmLayer(nn.Module):
    def __init__(self, input_size: int, hidden_size: int) -> None:
        super().__init__()
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.weight_ih = nn.Parameter(torch.empty(4 * hidden_size, input_size))
        self.weight_hh = nn.Parameter(torch.empty(4 * hidden_size, hidden_size))
        self.bias_ih = nn.Parameter(torch.empty(4 * hidden_size))
        self.bias_hh = nn.Parameter(torch.empty(4 * hidden_size))
        stdv = hidden_size**-0.5
        for p in self.parameters():
            nn.init.uniform_(p, -stdv, stdv)

    def forward(self, x: Tensor, state: tuple[Tensor, Tensor]) -> tuple[Tensor, tuple[Tensor, Tensor]]:
        h_prev, c_prev = state
        H = self.hidden_size
        gates_x = x @ self.weight_ih.t() + self.bias_ih
        gates_h = h_prev @ self.weight_hh.t() + self.bias_hh
        gates = gates_x + gates_h
        i = torch.sigmoid(gates[:, 0 * H : 1 * H])
        f = torch.sigmoid(gates[:, 1 * H : 2 * H])
        g = torch.tanh(gates[:, 2 * H : 3 * H])
        o = torch.sigmoid(gates[:, 3 * H : 4 * H])
        c_new = f * c_prev + i * g
        h_new = o * torch.tanh(c_new)
        return h_new, (h_new, c_new)

    def new_state(self, batch_size: int, device: Any | None = None) -> tuple[Tensor, Tensor]:
        # dtype tracks the parameter dtype so policy.double() / .float() propagates;
        # device defaults to the parameter's device so policy.to(device) propagates
        # naturally (torch.zeros(..., device=None) would silently fall back to CPU).
        target_device = device if device is not None else self.weight_ih.device
        zeros = torch.zeros(batch_size, self.hidden_size, device=target_device, dtype=self.weight_ih.dtype)
        return (zeros, zeros.clone())

    def to_flat(self) -> np.ndarray:
        """Canonical flat order: weight_ih row-major, weight_hh row-major, bias_ih, bias_hh.

        Matches Rust `LayerWeights for LstmLayer::to_flat` in
        src/rust/src/data/neural.rs.
        """
        return np.concatenate(
            [
                self.weight_ih.detach().cpu().numpy().astype(np.float64).ravel(),
                self.weight_hh.detach().cpu().numpy().astype(np.float64).ravel(),
                self.bias_ih.detach().cpu().numpy().astype(np.float64),
                self.bias_hh.detach().cpu().numpy().astype(np.float64),
            ]
        )

    def from_flat(self, slab: np.ndarray) -> None:
        """Load a flat slab in-place: weight_ih row-major, weight_hh row-major, bias_ih, bias_hh.

        Inverse of to_flat(); mirrors Rust `LayerWeights for LstmLayer::from_flat`.
        The full bias_ih slice is written verbatim -- init_v2_population already places
        the Jozefowicz +1.0 forget-bias in the correct slot; no per-gate splicing needed here.
        """
        four_h = 4 * self.hidden_size
        n_w_ih = four_h * self.input_size
        n_w_hh = four_h * self.hidden_size
        c = 0

        def _copy(param: torch.nn.Parameter, src: np.ndarray) -> None:
            param.copy_(torch.from_numpy(np.ascontiguousarray(src)).to(param.dtype))

        with torch.no_grad():
            _copy(self.weight_ih, slab[c : c + n_w_ih].reshape(four_h, self.input_size))
            c += n_w_ih
            _copy(self.weight_hh, slab[c : c + n_w_hh].reshape(four_h, self.hidden_size))
            c += n_w_hh
            _copy(self.bias_ih, slab[c : c + four_h])
            c += four_h
            _copy(self.bias_hh, slab[c : c + four_h])

    def extra_repr(self) -> str:
        return f"input_size={self.input_size}, hidden_size={self.hidden_size}"

File: src/rust/src/data/neural/layers/lstm.rs
//! LSTM cell layer (PyTorch nn.LSTMCell convention).

use super::super::Activation;
use super::helpers::dot_plus_bias;
use super::tensor::tensor_table;

/// LSTM cell matching PyTorch nn.LSTMCell convention (two biases, no peepholes).
///
/// Forward equations with gate ordering (i, f, g, o):
///   i_t = sigmoid(W_ii @ x_t + b_ii + W_hi @ h_{t-1} + b_hi)
///   f_t = sigmoid(W_if @ x_t + b_if + W_hf @ h_{t-1} + b_hf)
///   g_t =    tanh(W_ig @ x_t + b_ig + W_hg @ h_{t-1} + b_hg)
///   o_t = sigmoid(W_io @ x_t + b_io + W_ho @ h_{t-1} + b_ho)
///   c_t = f_t * c_{t-1} + i_t * g_t
///   h_t = o_t * tanh(c_t)
///
/// Weight storage matches torch.nn.LSTMCell:
///   weight_ih: [4H, input_size] with rows 0..H = W_ii, H..2H = W_if, 2H..3H = W_ig, 3H..4H = W_io
///   weight_hh: [4H, H]          with rows 0..H = W_hi, H..2H = W_hf, 2H..3H = W_hg, 3H..4H = W_ho
///   bias_ih:   [4H] in order b_ii, b_if, b_ig, b_io
///   bias_hh:   [4H] in order b_hi, b_hf, b_hg, b_ho
#[derive(Debug, Clone)]
pub struct LstmLayer {
    pub input_size: usize,
    pub hidden_size: usize,
    pub weight_ih: Vec<Vec<f64>>,
    pub weight_hh: Vec<Vec<f64>>,
    pub bias_ih: Vec<f64>,
    pub bias_hh: Vec<f64>,
}

impl LstmLayer {
    pub fn zeros(input_size: usize, hidden_size: usize) -> Self {
        let four_h = 4 * hidden_size;
        Self {
            input_size,
            hidden_size,
            weight_ih: vec![vec![0.0; input_size]; four_h],
            weight_hh: vec![vec![0.0; hidden_size]; four_h],
            bias_ih: vec![0.0; four_h],
            bias_hh: vec![0.0; four_h],
        }
    }

    /// Compute one forward step: (h_prev, c_prev, x) -> (h_new, c_new).
    pub fn forward(&self, h_prev: &[f64], c_prev: &[f64], x: &[f64]) -> (Vec<f64>, Vec<f64>) {
        assert_eq!(h_prev.len(), self.hidden_size);
        assert_eq!(c_prev.len(), self.hidden_size);
        assert_eq!(x.len(), self.input_size);
        let h = self.hidden_size;
        let mut h_new = vec![0.0; h];
        let mut c_new = vec![0.0; h];

        for idx in 0..h {
            // i gate: row idx
            let i = Activation::Sigmoid.apply(
                dot_plus_bias(&self.weight_ih[idx], x, self.bias_ih[idx])
                    + dot_plus_bias(&self.weight_hh[idx], h_prev, self.bias_hh[idx]),
            );
            // f gate: row idx + H
            let f = Activation::Sigmoid.apply(
                dot_plus_bias(&self.weight_ih[idx + h], x, self.bias_ih[idx + h])
                    + dot_plus_bias(&self.weight_hh[idx + h], h_prev, self.bias_hh[idx + h]),
            );
            // g gate (tanh, the "cell candidate"): row idx + 2H
            let g = (dot_plus_bias(&self.weight_ih[idx + 2 * h], x, self.bias_ih[idx + 2 * h])
                + dot_plus_bias(
                    &self.weight_hh[idx + 2 * h],
                    h_prev,
                    self.bias_hh[idx + 2 * h],
                ))
            .tanh();
            // o gate: row idx + 3H
            let o = Activation::Sigmoid.apply(
                dot_plus_bias(&self.weight_ih[idx + 3 * h], x, self.bias_ih[idx + 3 * h])
                    + dot_plus_bias(
                        &self.weight_hh[idx + 3 * h],
                        h_prev,
                        self.bias_hh[idx + 3 * h],
                    ),
            );

            c_new[idx] = f * c_prev[idx] + i * g;
            h_new[idx] = o * c_new[idx].tanh();
        }
        (h_new, c_new)
    }
}

// Flat order: weight_ih (row-major) -> weight_hh (row-major) -> bias_ih -> bias_hh.
tensor_table!(LstmLayer {
    weight_ih,
    weight_hh,
    bias_ih,
    bias_hh
});
